### PR TITLE
signal-desktop: init darwin at 7.5.1

### PR DESCRIPTION
**Pull request by juuyokka**
## Description of changes

Add support for x86_64-darwin and aarch64-darwin. I unfortunately could not figure out whether or not the update script works. I also could not figure out how to run the tests. If someone could tell me how to do both of those things I'd be happy to test those as well.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [X] `sandbox = relaxed`
  - [X] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by DontEatOreo** [comment]
Failed to pass `nixpkgs-review`

```console
nix run nixpkgs#nixpkgs-review -- pr 286188

$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/286188/head:refs/nixpkgs-review/1
remote: Enumerating objects: 12, done.
remote: Counting objects: 100% (7/7), done.
remote: Total 12 (delta 7), reused 7 (delta 7), pack-reused 5
Unpacking objects: 100% (12/12), 3.21 KiB | 235.00 KiB/s, done.
From https://github.com/NixOS/nixpkgs
 + 470882506cac...bf576e6b59b9 master                -> refs/nixpkgs-review/0  (forced update)
 + 3aee8feaaa63...89b5d0b40b61 refs/pull/286188/head -> refs/nixpkgs-review/1  (forced update)
$ git worktree add /Users/DontEatOreo/.cache/nixpkgs-review/pr-286188/nixpkgs bf576e6b59b9b5d46bf0e32d37cd437a093915c5
Preparing worktree (detached HEAD bf576e6b59b9)
Updating files: 100% (39380/39380), done.
HEAD is now at bf576e6b59b9 Merge pull request #285927 from presto8/home-assistant-emporia-vue
$ git merge --no-commit --no-ff 89b5d0b40b61c6dcf9a75b4532b7e229b3a9a24a
Automatic merge went well; stopped before committing as requested
$ nix build --nix-path nixpkgs=/Users/DontEatOreo/.cache/nixpkgs-review/pr-286188/nixpkgs nixpkgs-overlays=/tmp/tmph0u6d0gk --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation -f /Users/DontEatOreo/.cache/nixpkgs-review/pr-286188/build.nix
error: builder for '/nix/store/ib12i056ni9snibi3k5968z754hrl25m-signal-desktop-beta-6.47.0-beta.1.drv' failed with exit code 1;
       last 3 log lines:
       > Running phase: unpackPhase
       > error: readUDIFResourceFile - signature incorrect
       > error: Undefined error: 0
       For full logs, run 'nix log /nix/store/ib12i056ni9snibi3k5968z754hrl25m-signal-desktop-beta-6.47.0-beta.1.drv'.
error: 1 dependencies of derivation '/nix/store/12xmcvm1x889fj1f5i9b244gbk2qzrs9-review-shell.drv' failed to build

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/286188

1 package failed to build:
signal-desktop-beta

1 package built:
signal-desktop

$ /nix/store/ffkg7rz4zxfsdix6xxmhk2v3nx76r141-nix-2.18.1/bin/nix-shell /Users/DontEatOreo/.cache/nixpkgs-review/pr-286188/shell.nix --nix-path nixpkgs=/Users/DontEatOreo/.cache/nixpkgs-review/pr-286188/nixpkgs nixpkgs-overlays=/tmp/tmph0u6d0gk

[nix-shell:~/.cache/nixpkgs-review/pr-286188]$
```

`nix log /nix/store/ib12i056ni9snibi3k5968z754hrl25m-signal-desktop-beta-6.47.0-beta.1.drv`:
```
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
error: readUDIFResourceFile - signature incorrect
error: Undefined error: 0
```

**Comment by Mic92** [comment]
You can test the update script by first installing `nix-update` and than running this command from the root of the repository:

```
$ nix-update --use-update-script --build --commit signal-desktop
```

**Comment by juuyokka** [comment]
Checking back in on this PR, I have rebased my changes onto master and did a small refactor. I have also tested the update script to ensure it works.

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix [hunk 1]
```diff
@@ -1,6 +1,9 @@
 { hostPlatform, callPackage }: {
-  signal-desktop = if hostPlatform.system == "aarch64-linux"
-    then callPackage ./signal-desktop-aarch64.nix { }
-    else callPackage ./signal-desktop.nix { };
+  signal-desktop = {
+    aarch64-linux = callPackage ./signal-desktop-aarch64.nix { };
+    x86_64-linux = callPackage ./signal-desktop.nix { };
+    aarch64-darwin = callPackage ./signal-desktop-aarch64-darwin.nix { };
+    x86_64-darwin = callPackage ./signal-desktop-darwin.nix { };
+  }.${hostPlatform.system};
```
Comment (RossComputerGuy) [inline pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:7]: Shouldn't this be `targetPlatform`?

Comment (juuyokka) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:7]: The package originally used `hostPlatform`, so I just copied that into my PR. If `targetPlatform` is more favorable I can use that instead.

Comment (RossComputerGuy) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:7]: Yeah, please change it to use the target platform. You only use host when you need to use something that runs on the host platform (example: compilers, linkers, build tools, etc.)

Comment (DontEatOreo) [inline pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:7]: Wouldn't it be better to rename `signal-desktop.nix` to `signal-desktop.nix-x86_64-linux.nix` like the other files, so we can simplify it?
```suggestion
signal-desktop = callPackage ./"signal-desktop-${hostPlatform.system}.nix" { };
```

Comment (RossComputerGuy) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:7]: Agreed, a good example of this with a package I am a maintainer of is Lens.

Comment (juuyokka) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix]: Done. I also combined this with @RossComputerGuy's suggestion to change `hostPlatform` to `targetPlatform`.

**Comment by DontEatOreo** [comment]
Please rebase and reword your commit to `signal-desktop: refactor file names and targetPlatform`

**Comment by juuyokka** [comment]
Ah, that's my bad. Nothing a good 'ol `git commit --amend` can't fix!

**Review by 4evy — APPROVED** [review]

**Review by RossComputerGuy — APPROVED** [review]
image0

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix [hunk 2]
```diff
@@ -1,6 +1,4 @@
-{ hostPlatform, callPackage }: {
-  signal-desktop = if hostPlatform.system == "aarch64-linux"
-    then callPackage ./signal-desktop-aarch64.nix { }
-    else callPackage ./signal-desktop.nix { };
-  signal-desktop-beta = callPackage ./signal-desktop-beta.nix{ };
+{ targetPlatform, callPackage }: {
+  signal-desktop = callPackage ./signal-desktop-${targetPlatform.system}.nix { };
```
Comment (Mic92) [inline pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: Mhm. This will cause eval errors in case we have a new unsupported platform.
This needs to be fixed.

Comment (Mic92) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: Maybe builtins.pathExists with a fallback to x86_64-linux? But otherwise the previous if statement was also good enough, easier to read than some other complex logic.

Comment (juuyokka) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: Perhaps we could do something like this?
```suggestion
  signal-desktop =
    let
      default = callPackage ./signal-desktop-x86_64-linux.nix { };
      supportedPlatforms = default.meta.platforms;
      inherit (targetPlatform) system;
    in
    if builtins.elem system supportedPlatforms
    then callPackage ./signal-desktop-${system}.nix { }
    else default;
```
Or we could use `builtins.pathExists` like this
```suggestion
  signal-desktop =
    let
      path = ./signal-desktop-${targetPlatform.system}.nix;
      default = ./signal-desktop-x86_64-linux.nix;
    in
    if builtins.pathExists path
    then callPackage path { }
    else callPackage default { };
```
I think I prefer the latter. Evaluating `meta` in the first one would work but it's harder to follow.

**Comment by DontEatOreo** [comment]
Looking again at the commits, you're breaking the [Commit Conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).

> If you have commits `pkg-name: oh, forgot to insert whitespace`: squash commits in this case. Use `git rebase -i`. See [Squashing Commits](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_squashing) for additional information.

Just squash everything into first commit (89b5d0b40b61c6dcf9a75b4532b7e229b3a9a24a) and then reword it to `signal-desktop: init darwin at 7.5.1`

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix (hunk 2 above)
Comment (DontEatOreo) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: @Mic92 Wouldn't it be better to just `throw` and say your platform is not supported? Since there isn't any guarantee the `x86_64-linux` will work properly on a new platform either?

**Review by teutat3s — COMMENTED** [review]
Do we want to add the [beta version for darwin](https://support.signal.org/hc/en-us/articles/360007318471-Signal-Beta) as well?
https://updates.signal.org/desktop/signal-desktop-beta-mac-universal-7.9.0-beta.1.dmg

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix (hunk 2 above)
Comment (Mic92) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: I think a throw statement is not quite same as saying that the platform is not supported in `meta.platforms` semantically speaking.

Comment (Mic92) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: We already have quite a few eval errors here and we should not keep adding to it: https://hydra.nixos.org/jobset/nixpkgs/trunk#tabs-errors

Comment (DontEatOreo) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: > I think a throw statement is not quite same as saying that the platform is not supported in meta.platforms semantically speaking.

You're right, although what now? Do we default to building `x86_64-linux`? Wouldn't it be better to mark anything that isn't in `meta.platforms` as broken so we can entirely skip making a derivation and save resources? 

Also, if we were to default to building for `x86_64-linux`, wouldn't it be an issue if we could build it on another platform but not run it?

Comment (teutat3s) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: According to @wegank, `targePlatform` is only required when building a compiler https://github.com/NixOS/nixpkgs/pull/310053#discussion_r1594205907, see https://nix.dev/tutorials/cross-compilation.html#platforms.

Comment (RossComputerGuy) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: > and run the final executable on the target platform.

If `hostPlatform` were used and you were to cross compile architectures, wouldn't you end up with a binary that's incompatible if `hostPlatform` were used?

Comment (wegank) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:2]: The cross-compiled program, in that example the compiler, still runs on the host platform (and produces executables that run on the target platform).

**Comment by DontEatOreo** [comment]
@juuyokka Are you still interested in pursuing this PR?

**Comment by juuyokka** [comment]
@DontEatOreo apologies for the late reply, I mostly am lost on what else needs to be done (other than rebasing off of master) to get this merged.

**Comment by teutat3s** [comment]
Done in https://github.com/NixOS/nixpkgs/pull/348165
